jena-extras/jena-serviceenhancer/pom.xml

```diff
@@ -0,0 +1,185 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+   Licensed to the Apache Software Foundation (ASF) under one or more
+   contributor license agreements.  See the NOTICE file distributed with
+   this work for additional information regarding copyright ownership.
+   The ASF licenses this file to You under the Apache License, Version 2.0
+   (the "License"); you may not use this file except in compliance with
+   the License.  You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>jena-serviceenhancer</artifactId>
+  <name>Apache Jena - Extras - Service Enhancer</name>
+  <description>A plugin that extends the sparql SERVICE clauses with bulk requests, lateral joins and advanced result set caching.</description>
+
+  <parent>
+    <groupId>org.apache.jena</groupId>
+    <artifactId>jena-extras</artifactId>
+    <relativePath>..</relativePath>
+    <version>4.6.0-SNAPSHOT</version>
+  </parent>
+
+  <properties>
+    <automatic.module.name>org.apache.jena.serviceplugins</automatic.module.name>
+  </properties>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.apache.jena</groupId>
+      <artifactId>apache-jena-libs</artifactId>
+      <type>pom</type>
+    </dependency>
+
+    <dependency>
+      <groupId>com.github.ben-manes.caffeine</groupId>
```

Comment:
Switching to plain guava cache should be doable. The main benefit of caffeine is the async API with support for futures and cancellation. The main issue might be differences in atomicity of the callbacks.

Reply:
I can find a few Apache project using Caffine but they don't ship Apache-built binaries which is an additional level of problems with GPL2-CPE.  GPL2-CPE and Java is not a simple matter. (What is "linking"?)

Having something we can't include in a Fuseki binary is a high barrier to acceptance.

Caffine uses modules so (I think) excluding a jar isn't going to work if jena is used with modulepath.

This needs investigation. https://issues.apache.org/jira/browse/LEGAL-336.

What features of Caffine are used by this work? Does it make a difference? Given there is a lot of off-machine work going on, is the difference in caching going to be visible? Can we put this behind an interface? (CacheFactory)

(It would be great to get rid of the shaded Guava. Guava is more stable these days.)

Reply:
I added a cache implementation based on guava and removed the caffeine dependency. At least it succeeds on the test cases but maybe it introduced some issues we haven't spotted yet. I yet need to test it with our own project setup.

There is the AsyncClaimingCache interface and implementation that extends an undelying caching system with support for "claiming entries" and adding "eviction guards". Claimed entries are stored in a separate map as long as there are open claims. Eviction guards are simply java predicates that can 'catch' entries that are evicted by the underlying caching system and for which the predicate evaluates to true. Claims and guards both serve the purpose to temporarily bypass the underlying cache's eviction policy and both need to be closed eventually.